posydon/popsyn/synthetic_population.py

```diff
@@ -62,38 +68,61 @@ def __init__(self, path_to_ini, path_to_data=None, verbose=False, MODEL={}):
         if '.ini' not in path_to_ini:
             raise ValueError('You did not provide a valid path_to_ini!')
         else:
-            synthetic_pop_params = binarypop_kwargs_from_ini(path_to_ini)
-
-            self.metallicity = synthetic_pop_params['metallicity']
-
+            self.synthetic_pop_params = binarypop_kwargs_from_ini(path_to_ini)
+            self.metallicity = self.synthetic_pop_params['metallicity']
             if not isinstance( self.metallicity, list):
                 self.metallicity = [self.metallicity]
-
-            self.binary_populations = []
-            for met in self.metallicity[::-1]:
-                self.ini_kw = binarypop_kwargs_from_ini(path_to_ini)
-                self.ini_kw['metallicity'] = met
-                self.ini_kw['temp_directory'] = self.create_met_prefix(met) + self.ini_kw['temp_directory']
-                self.binary_populations.append(BinaryPopulation(**self.ini_kw))
-
-        if path_to_data is None:
-            return
-        elif (isinstance(path_to_data, list) and '.h5' in path_to_data[0]) or ('.h5' in path_to_data):
-            self.path_to_data = path_to_data
+            self.binary_populations = None
+        
+    def create_binary_populations(self):
+        """Create a list of BinaryPopulation objects."""
+        self.binary_populations = []
+        ini_kw = self.synthetic_pop_params.copy()
+        for met in self.metallicity[::-1]:
+            ini_kw['metallicity'] = met
+            ini_kw['temp_directory'] = self.create_met_prefix(met) + self.synthetic_pop_params['temp_directory']
+            self.binary_populations.append(BinaryPopulation(**ini_kw))
 
     def get_ini_kw(self):
-        return self.ini_kw
+        return self.synthetic_pop_params.copy()
 
     def evolve(self):
         """Evolve population(s) at given Z(s)."""
+        if self.binary_populations is None:
+            self.create_binary_populations()
         while self.binary_populations:
             pop =  self.binary_populations.pop()
             print( f'Z={pop.kwargs["metallicity"]:.2e} Z_sun' )
             pop.evolve()
-            met_prefix = f'{pop.kwargs["metallicity"]:.2e}_Zsun_'
-            pop.save( met_prefix + 'population.h5' )
```

Comment:
Why can we skip to save the outcome of the evolution before deleting it?

Reply:
Inside `BinaryPopulation.evolve()` the evolution is already saved for each individual process.

This code merged the individual subprocess files into one, which is now taken care of by `merge_parallel_runs()`

Reply:
Hm, why was it than saved here before?
Additionally, I'd suggest to rename the `evolve` function than to e.g. `evolve_and_save` to make clear, that this has to stay in this function.

Reply:
This `save` waited for all sub-processes and then called `combine_saved_files` to move the data from individual runs into a single population file. That part is now done by `merge_parallel_runs()`. 

`evolve` does save binaries to disk, but only during the run if it exceeds the ram limit, dump_rate, or a parallel run.